ghc-lib-gen/src/Ghclibgen.hs

```diff
@@ -408,6 +409,16 @@ applyPatchDisableCompileTimeOptimizations ghcFlavor =
           ("{-# OPTIONS_GHC -O0 #-}\n" ++)
           =<< readFile' file
 
+applyPatchGHCiMessage :: GhcFlavor -> IO ()
+applyPatchGHCiMessage ghcFlavor =
+  when (ghcFlavor == GhcMaster) $ do
+    writeFile messageHs .
+        replace
+          "#if MIN_VERSION_ghc_heap(8,11,0)"
+          "#ifndef MIN_VERSION_ghc_heap\n#define MIN_VERSION_ghc_heap(major1,major2,minor) ((major1) <  8 || (major1) == 8 && (major2) <  11 || (major1) == 8 && (major2) == 11 && (minor) <= 0)\n#endif /* MIN_VERSION_ghc_heap */\n#if MIN_VERSION_ghc_heap(8,11,0)"
```

Comment:
Wouldn't it be better to do it as described in https://gitlab.haskell.org/ghc/ghc/-/blob/master/docs/users_guide/phases.rst#L429?

Reply:
@hurryabit Can you be more explicit please? Even after reading that link I don't know what you are recommending.

Reply:
Perhaps you mean, write it like this?
```
        replace
          "#if MIN_VERSION_ghc_heap(8,11,0)"
          "#if !defined(MIN_VERSION_ghc_heap) || MIN_VERSION_ghc_heap(8,11,0)"
```
If so, I agree it's better and am testing it now.

Reply:
Perhaps you mean, write it like this?
```
        replace
          "#if MIN_VERSION_ghc_heap(8,11,0)"
          "#if !defined(MIN_VERSION_ghc_heap) || MIN_VERSION_ghc_heap(8,11,0)"
```
If so, ~I agree it's better and am testing it now~ unfortunately, that doesn't work. Another approach would be to delete the guard. I'm not convinced that's in any way better than what we have here.

Reply:
I meant replacing it with
```
#if defined(MIN_VERSION_ghc_heap)
#if MIN_VERSION_ghc_heap(8,11,0)
```
But I guess that's a bit cumbersome because we'd need two `#endif`s then. Replacing it with
```
#if defined(MIN_VERSION_ghc_heap) && MIN_VERSION_ghc_heap(8,11,0)
```
should do the job as well. The thing I'd like to avoid is redefining one of these `MIN_VERSION_xyz` macros.

Would that work?

Reply:
Normally, when building a program that depends on `ghc-heap`, you'd compile your program with `ghc -package ghc-heap ...`. GHC would then automatically produce a definition of `MIN_VERSION_ghc_heap`  for you that you can use in your program. In our case, we don't build against the `ghc-heap` package, rather we include its sources directly into our program. It follows that `MIN_VERSION_ghc_heap` is not generated and therefore not available to us.

Now, when we see

```
#if MIN_VERSION_ghc_heap(8,11,0)
instance Binary Heap.StgTSOProfInfo
instance Binary Heap.CostCentreStack
instance Binary Heap.CostCentre
instance Binary Heap.IndexTable
instance Binary Heap.WhatNext
instance Binary Heap.WhyBlocked
instance Binary Heap.TsoFlags
#endif
```

we want our program to behave as if `MIN_VERSION_ghc_heap(8,11,0)` holds and we get the guarded instance declarations. The way I've chosen to do that is (1) to synthesize the definition of `MIN_VERSION_ghc_heap` (after all, we are morally on that version we just don't have the macro definition) and inject it before the test. Another approach would be to (2) remove the guard and its corresponding `#endif` completely.

Your suggestion of 
```
#if defined(MIN_VERSION_ghc_heap)
#if MIN_VERSION_ghc_heap(8,11,0)
...
```
means that we won't get the definitions guarded by `MIN_VERSION_ghc_heap(8,11,0)` and the closely related alternative that the idea suggests  of saying `#if !defined(MIN_VERSION_ghc_heap) || MIN_VERSION_ghc_heap(8,11,0)` won't work because macros are fully expanded before they are evaluated so the bit after the `||` generates an error.

I can see other ways of getting the effect we want e.g. we can (3) generate this block (replicating the 8.11.0 case):

```
#if !defined(MIN_VERSION_ghc_heap)
instance Binary Heap.StgTSOProfInfo
instance Binary Heap.CostCentreStack
instance Binary Heap.CostCentre
instance Binary Heap.IndexTable
instance Binary Heap.WhatNext
instance Binary Heap.WhyBlocked
instance Binary Heap.TsoFlags
#endif
```

You might prefer this as it avoids us generating a (pseudo) definition for `MIN_VERSION_ghc_heap`? In any case I think our choices are basically one of (1), (2) or (3).

Reply:
Thanks for the detailed explanation. I wasn't aware that `MIN_VERSION_ghc_heap` is different from `MIN_VERSION_GLASGOW_HASKELL`.

Doesn't what the PR is doing so far amount to hardcoding the version of `ghc-heap` to 8.11.0? Shouldn't we instead replace with
```
#ifndef MIN_VERSION_ghc_heap
#define MIN_VERSION_ghc_heap(major1,major2,minor) ((major1) <  X || (major1) == X && (major2) <  Y || (major1) == X && (major2) == Y && (minor) <= Z)
#endif /* MIN_VERSION_ghc_heap */
#if MIN_VERSION_ghc_heap(8,11,0)
```
where `X.Y.Z` is the version of `ghc-heap` whose files we're copying? Can we actually get our hands on that version number? (Wrapping my head around these `MIN_VERSION_...` macros has just made it explode... 🤯)

Reply:
> Doesn't what the PR is doing so far amount to hardcoding the version of ghc-heap to 8.11.0?

Yes.

> Shouldn't we instead replace with
> [...]
> where X.Y.Z is the version of ghc-heap whose files we're copying?

That would be an improvement.

> Can we actually get our hands on that version number? 

Yes. I've reached the conclusion that`ghc-heap` (`ghc-boot`, `ghc-boot-th`, `ghci`) version numbers are coincident with the GHC version number determined by `./configure` and the contents of `configure.ac` (makes sense in retrospect). Accordingly, we can get the prevailing `X.Y.Z` by inspection of the contents of the `ghcversion.h` we generate. I'll push an update to this PR now that does that.

Reply:
If we work under the assumption that `ghc-heap` has the same version as GHC, can't we just replace `MIN_VERSION_ghc_heap` by `MIN_VERSION_GLASGOW_HASKELL` then?

Reply:
@hurryabit  I expect you mean `__GLASGOW_HASKELL__` and writing `MIN_VERSION_ghc_heap` in terms of that instead of reading `__GLASGOW_HASKELL__` from `ghc_version.h` as my patch currently does?

If so, no because the prevailing version of `__GLASGOW_HASKELL__` is that of the compiler being used to build ghc-lib, not the version of GHC sources being built.

Reply:
That's a very good point. Thanks for catching my bad reasoning.